SDK/UnityAudioPluginInterface.cpp

```diff
@@ -0,0 +1,220 @@
+    
+    effect_data->plugin->reset();
+    
+    return UNITY_AUDIODSP_OK;
+}
+
+static UNITY_AUDIODSP_RESULT UNITY_AUDIODSP_CALLBACK set_position_callback
+    (
+        UnityAudioEffectState * state,
+        unsigned int position
+    )
+{
+    auto effect_data = state->GetEffectData<audio::plugin::raw::EffectData>();
+    
+    effect_data->plugin->set_position(static_cast<int>(position));
+    
+    return UNITY_AUDIODSP_OK;
+}
+
+void declare_effect
+    (
+        UnityAudioEffectDefinition & definition,
+        audio::plugin::raw::RawEntryPoint entry_point
+    )
+{
+    memset(&definition, 0, sizeof(definition));
+    memcpy(definition.name, entry_point.name, sizeof(definition.name));
+    definition.structsize = sizeof(UnityAudioEffectDefinition);
+    definition.paramstructsize = sizeof(UnityAudioParameterDefinition);
+    definition.apiversion = UNITY_AUDIO_PLUGIN_API_VERSION;
+    definition.pluginversion = 0x010000;
+    definition.create = entry_point.raw_plugin_factory;
+    definition.release = release_callback;
+    definition.process = process_callback;
+    definition.reset = reset_callback;
+    definition.setfloatparameter = set_float_parameter_callback;
+    definition.getfloatparameter = get_float_parameter_callback;
+    definition.getfloatbuffer = get_float_buffer_callback;
+    definition.setposition = set_position_callback;
+    
+    auto parameters = entry_point.parameters();
+    
+    definition.numparameters = static_cast<UInt32>(parameters.size());
+    definition.paramdefs =
+        new UnityAudioParameterDefinition [definition.numparameters];
+    
+    for (auto & parameter_dsc : parameters)
+    {
+        auto index = &parameter_dsc - &parameters[0];
+        
+        memcpy(
+            definition.paramdefs[index].name,
+            parameter_dsc.name,
+            sizeof(UnityAudioParameterDefinition::name)
+        );
+        
+        memcpy(
+            definition.paramdefs[index].unit,
+            parameter_dsc.unit,
+            sizeof(UnityAudioParameterDefinition::unit)
+        );
+        
+        definition.paramdefs[index].description = "";
+        definition.paramdefs[index].defaultval = parameter_dsc.default_value;
+        definition.paramdefs[index].displayscale = parameter_dsc.display_scale;
+        definition.paramdefs[index].displayexponent = parameter_dsc.display_exp;
+        definition.paramdefs[index].min = parameter_dsc.min_value;
+        definition.paramdefs[index].max = parameter_dsc.max_value;
+    }
+}
+
+extern "C" UNITY_AUDIODSP_EXPORT_API int AUDIO_CALLING_CONVENTION
+    UnityGetAudioEffectDefinitions
+    (
+        UnityAudioEffectDefinition * * * output
+    )
+{
+    static std::atomic_flag is_initialised {};
+    static UnityAudioEffectDefinition definitions [256];
+    static UnityAudioEffectDefinition * definition_pointers [256];
+    
+    static const size_t entry_points_count = audio::entry_points.size();
+    
+    if (!is_initialised.test_and_set())
```

Comment:
Is this called concurrently?

Reply:
No, probably not. But the function might be called multiple times. An atomic_flag is probably overkill -- although it doesn't harm.